Add ONNX support for MarianMT models

lewtun opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

This PR adds support to export MarianMT models in the ONNX format. The underlying logic builds on the awesome refactor / feature enhancement that @michaelbenayoun has implemented in #14358 & #14700 - ~we should rebase this branch on `master` once that PR is merged to simplify the diff in this PR.~ (Done)

Currently, this PR supports ONNX exports for the following "tasks" (i.e. uses):

* `default`, `default-with-past` => equivalent to exporting a pretrained `MarianModel`
* `seq2seq-lm`, `seq2seq-lm-with-past` => equivalent to exporting a pretrained `MarianMTModel`
* `causal-lm`, `causal-lm-with-past`=> equivalent to exporting a pretrained `MarianForCausalLM`

Note that in each case, the end user will have to implement their own `generate()` method with the ONNX model - see [this BART example](https://github.com/huggingface/transformers/tree/master/examples/onnx/pytorch/summarization) for what's involved.

I've also checked locally that the "slow" tests pass with:

```
RUN_SLOW=1 pytest tests/test_onnx_v2.py -k "marian" -rp
```

## Usage

Here's a quick example to show how this works:

```python
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
from transformers.models.marian import MarianOnnxConfig

model_ckpt = "Helsinki-NLP/opus-mt-en-de"
tokenizer = AutoTokenizer.from_pretrained(model_ckpt)
ref_model = AutoModelForSeq2SeqLM.from_pretrained(model_ckpt)
# Export model
feature = "seq2seq-lm"
onnx_path = f"onnx/{model_ckpt}-{feature}/"
# Run this from a Jupyter notebook
!python -m transformers.onnx --model={model_ckpt} --atol=1e-4 --feature={feature} {onnx_path}
# Test export with inputs
batch_size = 4
encoder_inputs = tokenizer(
    ["Studies have been shown that owning a dog is good for you"] * batch_size,
    return_tensors="np",
)
decoder_inputs = tokenizer(
    ["Studien haben gezeigt dass es hilfreich ist einen Hund zu besitzen"]
    * batch_size,
    return_tensors="np",
)
all_inputs = {
    "input_ids": encoder_inputs["input_ids"],
    "attention_mask": encoder_inputs["attention_mask"],
    "decoder_input_ids": decoder_inputs["input_ids"],
    "decoder_attention_mask": decoder_inputs["attention_mask"],
}
# Generate ONNX outputs
ort_session = ort.InferenceSession(f"{onnx_path}model.onnx")
onnx_config = MarianOnnxConfig(ref_model.config, task=feature)
onnx_named_outputs = list(onnx_config.outputs.keys())
onnx_outputs = ort_session.run(onnx_named_outputs, all_inputs)
```

**TODO**

- [x] Extend support for language modelling head
- [x] Investigate range of numerical tolerance between raw and ONNX models for a range of checkpoints
- [x] Ensure that ONNX models are compatible with ONNX Runtime
- [x] Verify whether past key values are supported

Closes #13823, #13854 




## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

Maxinho96 committed:
First commit to add MarianMT to ONNX

Maxinho96 committed:
Now MarianModel.forward() automatically generates decoder_input_ids, like BartModel.forward()

Maxinho96 committed:
Adjusted MarianOnnxConfig.inputs and outputs to work with seq2seq-lm feature

Maxinho96 committed:
Style fix

michaelbenayoun committed:
Added support for other features for already supported models

michaelbenayoun committed:
Partial support for causal and seq2seq models

michaelbenayoun committed:
Partial support for causal and seq2seq models

lewtun committed:
Add default task for MarianMT ONNX

lewtun committed:
Remove automatic creation of decoder_input_ids

lewtun committed:
Extend inputs and outputs for MarianMT ONNX config

lewtun committed:
Add MarianMT to ONNX unit tests

lewtun committed:
Refactor

lewtun committed:
Merge branch 'master' into add-onnx-marianmt

michaelbenayoun committed:
OnnxSeq2SeqConfigWithPast to support seq2seq models

michaelbenayoun committed:
Parameterized the onnx tests

michaelbenayoun committed:
Restored run_mlm.py

michaelbenayoun committed:
Restored run_mlm.py

michaelbenayoun committed:
[WIP] BART update

michaelbenayoun committed:
BART and MBART

lewtun committed:
Merge branch 'onnx_enable_tasks_for_supported_models' into add-onnx-marianmt

lewtun committed:
Add past_key_values and fix dummy decoder inputs

Using a sequence length of 1 in generate_dummy_outputs() produces large discrepancies, presumably due to some hidden optimisations.

lewtun committed:
Refactor MarianOnnxConfig to remove custom past_key_values logic

lewtun committed:
Fix quality

lewtun commented:
There seems to be some sort of race condition happening in `run_tests_torch`:

```
_____________________________ ERROR collecting gw1 _____________________________
Different tests were collected between gw0 and gw1. The difference is:
--- gw0

+++ gw1
```

[This issue](https://github.com/pytest-dev/pytest-xdist/issues/366) has similar problems - perhaps a solution lies there.

michaelbenayoun committed:
Revert "Revert "Added support for other features for already supported models (#14358)" (#14679)"

This reverts commit 0f4e39c5599523c110cd713f60a3bfa145dad807.

michaelbenayoun committed:
Revert "Revert "Added support for other features for already supported models (#14358)" (#14679)"

This reverts commit 0f4e39c5599523c110cd713f60a3bfa145dad807.

michaelbenayoun committed:
Revert "Revert "Added support for other features for already supported models (#14358)" (#14679)"

This reverts commit 0f4e39c5599523c110cd713f60a3bfa145dad807.

michaelbenayoun committed:
is_torch_available test to avoid failing imports

michaelbenayoun committed:
is_torch_available test to avoid failing imports

michaelbenayoun committed:
is_torch_available test to avoid failing imports

michaelbenayoun committed:
sorting parameterize parameters to solve ERROR gw0 gw1

michaelbenayoun committed:
sorting parameterize parameters to solve ERROR gw0 gw1

michaelbenayoun committed:
sorting parameterize parameters to solve ERROR gw0 gw1

michaelbenayoun committed:
tests fix

michaelbenayoun committed:
tests fix

michaelbenayoun committed:
tests fix

michaelbenayoun committed:
tests fix

michaelbenayoun committed:
tests fix

michaelbenayoun committed:
tests fix

michaelbenayoun committed:
GPT2 with past fix

michaelbenayoun committed:
GPT2 with past fix

michaelbenayoun committed:
GPT2 with past fix

michaelbenayoun committed:
Fixed stateful class attribute change that was breaking things when converting multiple models sequentially

michaelbenayoun committed:
Fixed stateful class attribute change that was breaking things when converting multiple models sequentially

michaelbenayoun committed:
Fixed stateful class attribute change that was breaking things when converting multiple models sequentially

michaelbenayoun committed:
Removed onnx file

michaelbenayoun committed:
Removed onnx file

michaelbenayoun committed:
Removed onnx file

lewtun committed:
Merge branch 'onnx_enable_tasks_for_supported_models_part_2' into add-onnx-marianmt

lewtun committed:
Refactor Marian export to account for base changes

lewtun committed:
Fix copies

michaelbenayoun committed:
Implemented suggestions

michaelbenayoun committed:
Implemented suggestions

michaelbenayoun committed:
Implemented suggestions

lewtun committed:
Merge branch 'onnx_enable_tasks_for_supported_models_part_2' into add-onnx-marianmt

lewtun committed:
Extend support for causal LM

michaelbenayoun committed:
Fixed __init__ to resolve conflict with master

michaelbenayoun committed:
Fixed __init__ to resolve conflict with master

lewtun committed:
Merge branch 'onnx_enable_tasks_for_supported_models_part_2' into add-onnx-marianmt

michaelbenayoun committed:
Remove commented import

lewtun committed:
Merge branch 'onnx_enable_tasks_for_supported_models_part_2' into add-onnx-marianmt

lewtun committed:
Merge branch 'master' into add-onnx-marianmt

lewtun committed:
Remove ONNX model

lewtun committed:
Remove redundant class method

lewtun commented on docs/source/serialization.rst:
```diff
@@ -48,6 +48,7 @@ Ready-made configurations include the following models:
 - GPT Neo
 - LayoutLM
 - Longformer
+- Marian
```
I'm not sure whether `.rst` files are still allowed with the new `.mdx` doc - does this need updating / changing?

michaelbenayoun replied:
Letting @LysandreJik answering this one.

lewtun replied:
I saw the Sylvain recently converted all the RST files to MDX, so I'll rebase and this file should disappear :)

lewtun commented on src/transformers/models/marian/tokenization_marian.py:
```diff
@@ -310,7 +310,7 @@ def __setstate__(self, d: Dict) -> None:
         self._setup_normalizer()
 
-    def num_special_tokens_to_add(self, **unused):
+    def num_special_tokens_to_add(self, *args, **kwargs):
```
This change is required to accommodate the use of positional arguments like `tokenizer.num_special_tokens_to_add(is_pair)` in `_generate_dummy_inputs_for_sequence_classification_and_question_answering()`.

I'm not sure why we had `**unused` in the first place, but the change also seems more conventional IMO.

lewtun commented on src/transformers/models/marian/configuration_marian.py:
```diff
@@ -160,3 +166,226 @@ def __init__(
+            ]
+        return common_inputs
+
+    def _generate_dummy_inputs_for_sequence_classification_and_question_answering(
```
Technically, Marian doesn't have heads for sequence classification or question answering and this function is here due to the copy-paste from the BART config.

If you think this is confusing, I can remove this function and refactor the other dummy generation functions accordingly.

michaelbenayoun replied:
I think it can be done, you'll just have to remove the `# Copied from` comment at the top of the class declaration.

LysandreJik replied:
You could remove the `# Copied from` which is at the top of the class declaration and add it only to methods. It supports methods as well as classes.

lewtun committed:
Tidy up imports

lewtun committed:
Fix quality

lewtun commented on src/transformers/models/marian/configuration_marian.py:
```diff
@@ -160,3 +166,226 @@ def __init__(
         )
+
+
+# Copied from transformers.models.bart.configuration_bart.BartOnnxConfig with Bart->Marian
```
Since the Marian model is copied from BART (see `modeling_marian.py`), I adopted a similar approach for the ONNX config.

LysandreJik replied:
Yes, nice!

lewtun committed:
Merge branch 'master' into add-onnx-marianmt

LysandreJik reviewed (approved):
Looks good, thank you @lewtun!

LysandreJik commented:
Feel free to merge once you have taken care of the docs and the `# Copied from` statements :)

lewtun committed:
Refactor dummy input function

lewtun commented on src/transformers/models/marian/configuration_marian.py:
```diff
@@ -160,3 +166,221 @@ def __init__(
+            ]
+        return common_inputs
+
+    def _generate_dummy_inputs_for_encoder_and_decoder(
```
I renamed this function from `_generate_dummy_inputs_for_sequence_classification_and_question_answering()` to something that closer reflects its usage in the other dummy input functions.

lewtun committed:
Add copied from statements to Marian config functions

lewtun commented:
Thanks for the reviews @LysandreJik and @michaelbenayoun 🙏 !

I've fixed the docs by rebasing on `master` and added the `# Copied from:` snippets to the functions (I did not know about that trick!)

Will merge once all the test pass :)

lewtun committed:
Remove false copied from comments

lewtun committed:
Fix copy from comment

CatchDr commented:
The outputs decoding cannot get the correct result. How do you get the translation result
